### ISSUE TITLE
LoopVersioningLICM missing vectorization when loop bound is invariant under no-alias assumption

### ISSUE THREAD

**Issue by huihzhang** [llvm/llvm-project#112060]
There are some patterns I noticed from my previous customer codebase study:
t1.cpp
```
void foo(int *A, int *B, int *LoopBound) {
  for (int k = 0; k < *LoopBound; k++)
    A[k] += B[k];
}
```
t2.cpp
```
class Base {
protected:
  int m_totalPhase;
};
class Derived : public Base {
public:
  void foo(int* A, int *B);
};

void Derived::foo(int *A, int *B) {
  for (int k = 0; k < m_totalPhase; k++)
    A[k] += B[k];
}
```

Loop bound in both cases requires load from pointer that may alias with memory accesses in loop. 
```
clang++ --target=aarch64 -mcpu=cortex-a57 -c -O3 t[1|2].cpp -Rpass-missed=loop-vectorize

t1.cpp:2:3: remark: loop not vectorized [-Rpass-missed=loop-vectorize]
    2 |   for (int k = 0; k < *LoopBound; k++)
```

Polly is creating runtime versioning to achieve similar effect like loop versioning. The load of loop bound is hoisted outside, thus vectorization achieved.
```
clang++ --target=aarch64 -mcpu=cortex-a57 -c -O3 t[1|2].cpp -Rpass=loop-vectorize -mllvm -polly -mllvm -polly-invariant-load-hoisting -mllvm -polly-process-unprofitable

t1.cpp:2:3: remark: vectorized loop (vectorization width: 4, interleaved count: 4) [-Rpass=loop-vectorize]
    2 |   for (int k = 0; k < *LoopBound; k++)
t2.cpp:12:3: remark: vectorized loop (vectorization width: 4, interleaved count: 4) [-Rpass=loop-vectorize]
   12 |   for (int k = 0; k < m_totalPhase; k++)
```

Both cases are good fit for LoopVersioningLICM, where load of loop bound can be hoisted for the versioned loop with no-alias assumption.

I wonder if there are enough interest in community to support vectorization for such loops. I am currently working on different projects, so might not have enough bandwidth to proceed with upstream fixes.

**Comment by pinskia** [comment]
Note the corresponding GCC issue: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115363 .

**Comment by appujee** [comment]
This is the loop before loop-versioning-licm

```llvm
define dso_local void @foo(int*, int*, int*)(ptr nocapture noundef %A, ptr nocapture noundef readonly %B, ptr nocapture noundef readonly %LoopBound) local_unnamed_addr {
entry:
  %0 = load i32, ptr %LoopBound, align 4
  %cmp6 = icmp sgt i32 %0, 0
  br i1 %cmp6, label %for.body, label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.body, %entry
  ret void

for.body:                                         ; preds = %entry, %for.body
  %indvars.iv = phi i64 [ %indvars.iv.next, %for.body ], [ 0, %entry ]
  %arrayidx = getelementptr inbounds i32, ptr %B, i64 %indvars.iv
  %1 = load i32, ptr %arrayidx, align 4
  %arrayidx2 = getelementptr inbounds i32, ptr %A, i64 %indvars.iv
  %2 = load i32, ptr %arrayidx2, align 4
  %add = add nsw i32 %2, %1
  store i32 %add, ptr %arrayidx2, align 4
  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
  %3 = load i32, ptr %LoopBound, align 4
  %4 = sext i32 %3 to i64
  %cmp = icmp slt i64 %indvars.iv.next, %4
  br i1 %cmp, label %for.body, label %for.cond.cleanup
}
```

**Comment by appujee** [comment]
Adding restrict to the bound vectorizes it https://godbolt.org/z/bGE9EoWG3, so making the the loop versioning work for this case should trigger autovectorization.

**Comment by appujee** [comment]
In fact, it is not loop-versioning, as long as there is a way to  `restrict` the pointers the vectorization will happen. so something like

```cpp
#include<algorithm>

void foo(int *A, int *B, int *N) {
  int p1 = *N;
  if (std::abs(A-N) > p1) {
    for (int k = 0; k < p1; k++)
        A[k] += B[k];
  } else {
    for (int k = 0; k < *N; k++)
        A[k] += B[k];
  }
}
```

https://godbolt.org/z/b1xhf3bb7

**Comment by chrisj-quic** [comment]
It looks like LoopVersioningLICM bails out (LoopVersioningLICM::legalLoopStructure) as SCEV is unable to compute the loop trip count required to generate alias bounds checks.

**Comment by hiraditya** [comment]
cc: @ayalz

**Comment by hiraditya** [comment]
As per Alexey: loop-vectorizer already has predicated-scev analysis so this case needs to be added.

**Comment by pawan-nirpal-031** [comment]
Hand modifying the test to get rid of the load in the loop body, and replacing it's uses with the value of load from header allows the loop to be vectorized with runtime checks. Being able to prove that the loopbound load in the loop body is invariant and a hoist of that load allows the loop vectorizer to proceed.

Other observation, when we remove -polly-invariant-load-hoisting option, even polly fails to generate vector version of the loop. 

change.ll

```llvm
define dso_local void @_Z3fooPiS_S_(ptr noundef captures(none) %A, ptr noundef readonly captures(none) %B, ptr noundef readonly captures(none) %LoopBound) local_unnamed_addr #0 {
entry:
  %0 = load i32, ptr %LoopBound, align 4, !tbaa !5
  %cmp6 = icmp sgt i32 %0, 0
  br i1 %cmp6, label %for.body, label %for.cond.cleanup

for.cond.cleanup:                                 ; preds = %for.body, %entry
  ret void

for.body:                                         ; preds = %for.body, %entry
  %indvars.iv = phi i64 [ %indvars.iv.next, %for.body ], [ 0, %entry ]
  %arrayidx = getelementptr inbounds nuw i32, ptr %B, i64 %indvars.iv
  %1 = load i32, ptr %arrayidx, align 4, !tbaa !5
  %arrayidx2 = getelementptr inbounds nuw i32, ptr %A, i64 %indvars.iv
  %2 = load i32, ptr %arrayidx2, align 4, !tbaa !5
  %add = add nsw i32 %2, %1
  store i32 %add, ptr %arrayidx2, align 4, !tbaa !5
  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
  %4 = sext i32 %0 to i64
  %cmp = icmp slt i64 %indvars.iv.next, %4
  br i1 %cmp, label %for.body, label %for.cond.cleanup, !llvm.loop !9
}
```

@hiraditya, @fhahn @david-arm I am trying to approach this using a similar analysis as ```bool polly::isHoistableLoad``` Which I think helps polly prove the loop bound load is invariant, unless there's something fundamental from polly infra like ISL limiting us from replicating something like this in LVerLICM or just good old LICM as well we can get the loop to vectorize.

**Comment by fhahn** [comment]
@pawan-nirpal-031 are you saying `LVerLICM` is not hoisting the load? I would have expected it to already be able to do something like that.

**Comment by mk-srivastava** [comment]
Hello Everyone!

I have a PR ready for this issue, currently I am validating it. 

The PR handles this issue, by modifying the LVerLICM legality check specifically `SCEVCouldNotCompute` one. 

The approach is to find the loop's dynamic bound in question, and walk backwards from the dynamic bound to find the chain of instructions part of a loop that compute the dynamic bound. 

If all of these bound computing instructions are are invariant, safe to hoist. If the bound-computing instructions have load instructions in them, then as part of legality checks is to ensure they are simple(not volatile) and don't read from a pointer that is stored within in the loop. 

Once the above legality is confirmed about the dynamic bound and the instructions computing it, then we can safely version the loop with a deferred runtime check by giving it a placeholder condition in the preheader. 

Once the loop is versioned, then bound-computing instructions and the dynamic bound can be hoisted out of the versioned loop into the preheader, making the trip count computable by SCEVs. The placeholder then can be replaced with a actual runtime pointer overlap checks that verifies that the hoisted bound does not alias with any memory locations written to inside the loop. 

If aliasing is detected at runtime, execution falls back to the original unmodified loop(non-versioned loop).

**Comment by pawan-nirpal-031** [comment]
@mk-srivastava Thanks for your PR, However since LverLICM is turned off by default, we are trying to have this capability in Loop Vectorizer natively, Per me the idea is If such patterns can otherwise unblock vectorization via versioning, they need to be integrated in LV itself, there could be other such patterns, I'm hoping this turns out to be a motivating example for reattempting speculative vectorization via versioning. I also had the chance to try out your PR and I couldn't get vectorization after your patch has run. That's the motive behind this issue, Hoisting the load in itself may not be as beneficial. 

PS : I hope you're okay with me using some of your tests!

**Comment by mk-srivastava** [comment]
Hello @pawan-nirpal-031, 

> I couldn't get vectorization after your patch has run.

The patch will unblock loop-vectorizer for the loops with the dynamic bounds. You can try this by scheduling loop-vectorizer for the said loops after my patch, those loops should get vectorized with the changes. 

> Since LverLICM is turned off by default, we are trying to have this capability in Loop Vectorizer natively

The LoopVersioningLICM pass is practically designed to handle these kind of loops. So, I felt it will be better to implement this there for simplicity.

**Comment by pawan-nirpal-031** [comment]
@mk-srivastava Thanks for your reply, Our objective with this issue is to have the loop vectorizer natively do this instead of having to rely on LverLICM, besides LverLICM is not enabled by default, which will take this issue into a direction we don't hope for yet, it will be a separate project on it's own. The other reason is having this transform tightly coupled with LV will allow us to be mindful of the cases where such a transform is not profitable, we hope to levergae LV's cost models to determine if such a runtime check block might not introduce performance overheads, ie LV cost model aware rtcheck generation.

**Comment by nema-ashutosh** [comment]
Extending support for dynamic bounds to the loop-versioning utility would offer broader applicability compared to building local solutions within LVerLICM/LV. Has this approach been explored ?

**Comment by pawan-nirpal-031** [comment]
@nema-ashutosh Thanks for your comment. I haven't tried that, Reason being, for this particular issue, we would like such cases to be LV cost model aware, Hoisting the load speculatively may not always unblock vectorization, the intent is to try to unblock such cases ( and hopefully motivate more such re-attemp patterns of vectorization via versioning ) I believe leveraging this would be most beneficial, rather than trying to do the cost modeling in Loop Versioning or LVerLICM ( afaik there isn't much of a vectorization aware cost model in both these passes ). 

PS : I'm working towards a PR in LV https://github.com/llvm/llvm-project/pull/195027 and I see improved loop coverage in SPEC, need to weed out a last remaining few correctness issues, and this should be ready for full review. 

CC @efriedma-quic @chrisj-quic

**Comment by pawan-nirpal-031** [comment]
Hi @mk-srivastava please go ahead with your PR, the community came to a consensus it's best to do it out of the LV. Also add me as a reviewer, Thanks! :)

**Comment by mk-srivastava** [comment]
Hello @pawan-nirpal-031 ,

It seems I don't have the permission to add the reviewers. I will look into this.
